### PR TITLE
remove oracle withdraw and allow contract owner to withdraw

### PR DESCRIPTION
**Pull request by jinhoonbang**
- remove oracle withdraw and allow contract owner to withdraw
- open question. should there be an additional identity who can withdraw?
	- will increase contract size
	
	
### On-chain testing

```
$ forge test -vvv --match-path test/v0.8/foundry/vrf/VRFV2PlusSubscriptionAPI.t.sol
...
Test result: ok. 32 passed; 0 failed; 0 skipped; finished in 7.41ms
Ran 1 test suites: 32 tests passed, 0 failed, 0 skipped (32 total tests)
```

### Off-chain integration testing

```
$ cd integration-tests
$ make build_docker_image image=chainlink tag=local_dev
$ TEST_LOG_LEVEL=debug \
SELECTED_NETWORKS=SIMULATED \
CHAINLINK_IMAGE=chainlink \
CHAINLINK_VERSION=local_dev \
go test -v -count=1 -timeout 15m -run "^TestVRFv2Plus$" ./smoke
```

### CODE REVIEW

File: contracts/src/v0.8/vrf/dev/SubscriptionAPI.sol [hunk 1]
```diff
@@ -204,35 +204,35 @@ abstract contract SubscriptionAPI is ConfirmedOwner, IERC677Receiver, IVRFSubscr
...
    * @param recipient where to send the funds
    * @param amount amount to withdraw
    */
-  function oracleWithdraw(address recipient, uint96 amount) external nonReentrant {
+  function withdraw(address recipient, uint96 amount) external nonReentrant onlyOwner {
     if (address(LINK) == address(0)) {
       revert LinkNotSet();
     }
-    if (s_withdrawableTokens[msg.sender] < amount) {
+    if (s_withdrawableTokens < amount) {
```
Comment (kidambisrinivas) [inline contracts/src/v0.8/vrf/dev/SubscriptionAPI.sol:215]: The behaviour change here is that
- Previously, the oracles could withdraw the earnings from VRFCoordinator
- Now, only the contract owner can withdraw the earnings
Is this correct?

File: contracts/src/v0.8/vrf/dev/VRFCoordinatorV2_5.sol [hunk 2]
```diff
@@ -44,11 +44,11 @@ contract VRFCoordinatorV2_5 is VRF, SubscriptionAPI, IVRFCoordinatorV2Plus {
...
     bytes extraArgs;
   }
-  mapping(bytes32 => address) /* keyHash */ /* oracle */ public s_provingKeys;
+  mapping(bytes32 => bool) /* keyHash */ /* exists */ public s_provingKeys;
   bytes32[] public s_provingKeyHashes;
   mapping(uint256 => bytes32) /* requestID */ /* commitment */ public s_requestCommitments;
-  event ProvingKeyRegistered(bytes32 keyHash, address indexed oracle);
-  event ProvingKeyDeregistered(bytes32 keyHash, address indexed oracle);
+  event ProvingKeyRegistered(bytes32 keyHash);
+  event ProvingKeyDeregistered(bytes32 keyHash);
```
Comment (kidambisrinivas) [inline contracts/src/v0.8/vrf/dev/VRFCoordinatorV2_5.sol:51]: Now, proving keys are not registered to an oracle?

File: contracts/src/v0.8/vrf/dev/SubscriptionAPI.sol (hunk 1 above)
Comment (jinhoonbang) [reply contracts/src/v0.8/vrf/dev/SubscriptionAPI.sol:215]: yeah. "oracles" could have been any identity previously. It didn't necessarily need to be the node fulfilling requests

File: contracts/src/v0.8/vrf/dev/VRFCoordinatorV2_5.sol (hunk 2 above)
Comment (jinhoonbang) [reply contracts/src/v0.8/vrf/dev/VRFCoordinatorV2_5.sol:51]: yes correct

File: core/services/vrf/v2/integration_helpers_test.go [hunk 3]
```diff
@@ -1683,8 +1683,13 @@ func testMaliciousConsumer(
...
 	require.NoError(t, err)
-	_, err = uni.rootContract.RegisterProvingKey(
-		uni.neil, uni.nallory.From, pair(secp256k1.Coordinates(p)))
+	if vrfVersion == vrfcommon.V2Plus {
+		_, err = uni.rootContract.RegisterProvingKey(
+			uni.neil, nil, pair(secp256k1.Coordinates(p)))
+	} else {
+		_, err = uni.rootContract.RegisterProvingKey(
+			uni.neil, &uni.nallory.From, pair(secp256k1.Coordinates(p)))
+	}
```
Comment (dkneisly) [inline core/services/vrf/v2/integration_helpers_test.go:1692]: might be worth pulling this chunk into a separate method considering it's repeated 4 times

File: integration-tests/smoke/vrfv2plus_test.go [hunk 4]
```diff
@@ -525,7 +524,7 @@ func TestVRFv2Plus(t *testing.T) {
 	})
 	t.Run("Oracle Withdraw", func(t *testing.T) {
```
Comment (iljapavlovs) [inline integration-tests/smoke/vrfv2plus_test.go:525]: need to change test name

File: contracts/src/v0.8/vrf/dev/VRFCoordinatorV2_5.sol (hunk 2 above)
Comment (kidambisrinivas) [reply contracts/src/v0.8/vrf/dev/VRFCoordinatorV2_5.sol:51]: So, no matter how many proving keys we register, all earnings can only be withdrawn by contract owner now

File: contracts/scripts/native_solc_compile_all_vrfv2plus [hunk 5]
```diff
@@ -0,0 +1,60 @@
+#!/usr/bin/env bash
```
Comment (jinhoonbang) [inline contracts/scripts/native_solc_compile_all_vrfv2plus:1]: is this needed? this is already in https://github.com/smartcontractkit/chainlink/blob/develop/contracts/scripts/native_solc_compile_all_vrf

Comment (jinhoonbang) [reply contracts/scripts/native_solc_compile_all_vrfv2plus:1]: the naming convention here is defined [here](https://github.com/smartcontractkit/chainlink/blob/develop/contracts/GNUmakefile#L65) . I don't think maintaining a separate foundry profile (e.g. vrfv2plus) is needed

File: contracts/src/v0.8/vrf/dev/VRFCoordinatorV2_5.sol [hunk 6]
```diff
@@ -94,28 +94,26 @@ contract VRFCoordinatorV2_5 is VRF, SubscriptionAPI, IVRFCoordinatorV2Plus {
   }
 
   /**
-   * @notice Registers a proving key to an oracle.
-   * @param oracle address of the oracle
+   * @notice Registers a proving key to.
    * @param publicProvingKey key that oracle can use to submit vrf fulfillments
    */
-  function registerProvingKey(address oracle, uint256[2] calldata publicProvingKey) external onlyOwner {
+  function registerProvingKey(uint256[2] calldata publicProvingKey) external onlyOwner {
```
Comment (ibrajer) [inline contracts/src/v0.8/vrf/dev/VRFCoordinatorV2_5.sol:100]: @jinhoonbang @kidambisrinivas FYI this is a breaking change for Gauntlet commands (both for register and deregister proving key). When I use this branch, unit tests on the Gauntlet repo for proving key registration and deregistration are failing. I will fix those commands in the same PR where I would be introducing new withdrawal functions. Then, we would merge it on the Gauntlet develop branch only when this PR is merged (and after I double-check everything once again).

File: core/services/vrf/v2/integration_helpers_test.go (hunk 3 above)
Comment (jinhoonbang) [reply core/services/vrf/v2/integration_helpers_test.go:1692]: makes sense 👍

File: core/services/vrf/v2/coordinator_v2x_interface.go [hunk 7]
```diff
@@ -36,9 +36,10 @@ type CoordinatorV2_X interface {
 	GetConfig(opts *bind.CallOpts) (Config, error)
 	ParseLog(log types.Log) (generated.AbigenLog, error)
 	OracleWithdraw(opts *bind.TransactOpts, recipient common.Address, amount *big.Int) (*types.Transaction, error)
+	Withdraw(opts *bind.TransactOpts, recipient common.Address) (*types.Transaction, error)
```
Comment (dkneisly) [inline core/services/vrf/v2/coordinator_v2x_interface.go:39]: Do we need a second method here for withdrawing native?

Comment (jinhoonbang) [reply core/services/vrf/v2/coordinator_v2x_interface.go:39]: Yeah looks like `WithdrawNative()` could use a test 👍

Comment (jinhoonbang) [reply core/services/vrf/v2/coordinator_v2x_interface.go:39]: In terms of test coverage, withdrawNative is tested in integration tests which uses a different interface from this one.

Comment (ibrajer) [reply core/services/vrf/v2/coordinator_v2x_interface.go:39]: So this interface is not tested or is just the `Withdraw()` method not used/tested? And why do integration tests use a different interface?

Comment (jinhoonbang) [reply core/services/vrf/v2/coordinator_v2x_interface.go:39]: The interface is used and tested but seems that a few methods (`withdraw()` and `withdrawNative()`) are unused. 

The integration tests use different interface but I'm not sure if we have a good reason for it. I think we have room for some refactoring in integration tests to either create a shared interface for v2 and v2plus or use the existing one above.

**Review by ibrajer — APPROVED** [review]
I'm approving the PR to unblock you this week - but please take it with a grain of salt. Also, since this is impacting only v2.5 it should not break anything and would probably be okay to patch things next week (if needed).

**Pull request merged**